Added docs for 1498 (extending custom field filtering) (#2167)
nautobot/nautobot: nautobot/docs/rest-api/filtering.md

GET /api/dcim/sites/?cf_foo=123
```

!!! note

>     For custom field filters, due to historical details of implementation, only a single filter value can be specified when matching a given field. In other words, in the above example, you could *not* add `&cf_foo=456` to the query in order to get all sites where custom field `foo` is 123 *or* 456; instead you would need to run two separate queries. This restriction does not apply to custom field filters using lookup expressions (next section) and will likely be removed in a future major version of Nautobot.

Reviewer: ```suggestion
    For custom field filters, due to historical details of implementation, only a single filter value can be specified when matching a given field. In other words, in the above example, you could *not* add `&cf_foo=456` to the query in order to get all sites where custom field `foo` is 123 *or* 456; instead you would need to run two separate queries. This restriction does not apply to custom field filters using lookup expressions (next section) and will likely be removed in a future major version of Nautobot.

```

GET /api/dcim/sites/?cf_foo=123
```

!!! note

>     For custom field filters, due to historical details of implementation, only a single filter value can be specified when matching a given field. In other words, in the above example, you could _not_ add `&cf_foo=456` to the query in order to get all sites where custom field `foo` is 123 _or_ 456; instead you would need to run two separate queries. This restriction does not apply to custom field filters using lookup expressions (next section) and will likely be removed in a future major version of Nautobot.

Reviewer 2: ```suggestion
    For custom field filters, due to historical details of implementation, only a single filter value can be specified when matching a given field. In other words, in the above example, you could _not_ add `&cf_foo=456` to the query in order to get all sites where custom field `foo` is 123 _or_ 456; instead you would need to run two separate queries. This restriction does not apply to custom field filters using lookup expressions (next section) and will likely be changed in a future major version of Nautobot.
```

GET /api/dcim/sites/?cf_foo=123
```

> Multi-value filtering is available for all field types and lookup expressions. When multiple values are supplied, an `or` lookup will be performed by the query. For example, the following query will return only sites where a custom field named `foo` is equal to 123 _or_ 456:

Reviewer: I thought that in #2126, for now we have only single-value support for the pre-existing default/exact lookup filter on custom fields? That multi-value was only for the newly implemented lookup-expression cases below?

Reviewer 2: The default filter is still single value for backward compatibility. So for an `integer` custom field you would have these filters.

- Single value filter
  - `cf_cf1` (exact)
- Multi value filters
  - `cf_cf1_n` (exact negated)
  - `cf_cf1_lte` (less than or equal)
  - `cf_cf1_lt` (less than)
  - `cf_cf1_gte` (greater than or equal)
  - `cf_cf1_gt` (greater than)

One more thing to point out is not all custom field types got the extended lookup expression filters. We added extended filters for `text`, `url`, `select`, `integer` and `date` custom field types but not `boolean`, `multi-select` or `json`. We could add support for `multi-select` and `json` but those are going to require additional logic and were out of scope for the requested feature. I would just point out that the extended filters are supported in `text`, `url`, `select`, `integer` and `date` fields. The filters added are documented in the summary of #2126

Reviewer: So I feel like the example in this PR that's doing an exact match is still incorrect:

`GET /api/dcim/sites/?cf_foo=123&cf_foo=456`

because this is an exact match, it will *not* do a multi-value search - instead the last specified value will "win" as the single value to filter on, and the previous values will be ignored, i.e. this will only filter for `cf_foo=456`.

https://next.demo.nautobot.com/api/dcim/sites/?cf_site_type=POP&cf_site_type=NON_POP demonstrates this - it only shows sites whose `site_type` CF is set to `NON_POP`, not also those set to `POP`.

```no-highlight
GET /api/dcim/sites/?cf_foo=123&cf_foo=456
```

Custom fields can be mixed with built-in fields to further narrow results. When creating a custom string field, the type of filtering selected (loose versus exact) determines whether partial or full matching is used.

> Custom fields can use the lookup expressions listed in the next section by prepending `cf_` to the custom field name followed by the required lookup type (see below).

Reviewer: Hmm, this may be a gap/flaw with the changes in #2136 - after that PR, in API version 1.4+ the REST API representation of custom fields uses the `slug` rather than the `name` of the fields, i.e.:

```
"custom_fields": {
    "my_custom_field_slug": "value",
    "another_custom_field_slug": "value2"
}
```

versus the pre-1.4 API behavior of:

```
"custom_fields": {
    "My custom field!": "value",
    "Some / other custom field?": "value2"
}
```

So it would probably be better if the REST API filters in API version 1.4+ also used `slug` instead of `name`... but we don't have a way to do versioned FilterSets, so that would be a breaking change to existing filters. Probably the best we can do in 1.x is call out that regardless of REST API version the filtering of custom fields always uses the field name rather than the slug for now?

GET /api/dcim/sites/?cf_foo=123
```

> Multi-value filtering (as well as single-value filtering) is available for the `text`, `url`, `select`, `integer` and `date` field types and their associated lookup expressions. When multiple values are supplied, both filter lookups have to be satisfied.
> 
> In the following example, we have a custom field (named "Warranty Expires") with a type of `date`. The following query will return results whose "Warranty Expires" is on or after `2022-07-28` or before `2022-08-08`:
> 
> ```no-highlight
> GET /api/dcim/sites/?cf_warranty_expires__gte=2022-07-28&cf_warranty_expires__lt=2022-08-08
> ```
> 

Reviewer: ```suggestion
!!! note
    For custom field filters, due to historical details of implementation, only a single filter value can be specified when matching a given field. In other words, in the above example, you could *not* add `&cf_foo=456` to the query in order to get all sites where custom field `foo` is 123 *or* 456; instead you would need to run two separate queries. This restriction does not apply to custom field filters using lookup expressions (next section) and will likely be removed in a future major version of Nautobot.
```

Custom fields can be mixed with built-in fields to further narrow results. When creating a custom string field, the type of filtering selected (loose versus exact) determines whether partial or full matching is used.

Custom fields can use the lookup expressions listed in the next section by prepending `cf_` to the custom field `name` (and not the `slug`) followed by the required lookup type (see below).

## Lookup Expressions

> Certain model fields also support filtering using additional lookup expressions. This allows

Reviewer: ```suggestion
Certain model fields (including custom fields of type `text`, `url`, `select`, `integer`, and `date`) also support filtering using additional lookup expressions. This allows
```
